libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp

```diff
@@ -372,6 +402,7 @@ void NavEKF2_core::SelectVelPosFusion()
 
     // Check for data at the fusion time horizon
     extNavDataToFuse = storedExtNav.recall(extNavDataDelayed, imuDataDelayed.time_ms);
+    extNavVelToFuse = storedExtNavVel.recall(extNavVelDelayed, imuDataDelayed.time_ms);
```

Comment:
If the position offset correction is applied to extNavVelDelayed.vel here when extNavVelToFuse == true, then separate corrections in reset and fusion operations will not be necessary.

Reply:
Thank you @priseborough 
Maybe I can put position offset correction in writeExtNavVelData() inside AP_NavEKF2_Measurements.cpp ?

Reply:
Thanks for this.  So this comment has certainly uncovered **an existing bug** in the NavEKF2/3_core::ResetHeight() functions for **GPS** and we've just copied this bug for external nav.  When we reset the vertical velocity we are using an uncorrected velocity (i.e. sensor position correction has not been applied).

I'm obviously not the expert but I think @chobitsfan's solution to apply the correction inside writeExtNavVelData() means the correction will be using the incorrect imu data won't it?
CorrectExtNavVelForSensorOffset() has these two lines:

> const Vector3f angRate = imuDataDelayed.delAng * (1.0f/imuDataDelayed.delAngDT);
>  ext_velocity += get_vel_correction_for_sensor_offset(posOffsetBody, prevTnb, angRate);

I think Paul's original suggestion was to apply the correction to extNavVelDelayed when it is set using the recall() function but this leads to a slightly confusing situation where extNavVelDelayed is corrected but extNavVelNew is not.  We could handle this with some documentation or by changing the variable name but my feeling at the moment is we should just leave it as it was.

Reply:
I've raised issue https://github.com/ArduPilot/ardupilot/issues/14462 to cover the existing bug

Reply:
Thank you @rmackay9 

>  to apply the correction inside writeExtNavVelData() means the correction will be using the incorrect imu data 

Maybe we could change 
`const Vector3f angRate = imuDataDelayed.delAng * (1.0f/imuDataDelayed.delAngDT);`
in CorrectExtNavVelForSensorOffset() to
`const Vector3f angRate = imuDataNew.delAng * (1.0f/imuDataDelayed.delAngDT);`
New arrival ext nav vel data will be corrected using new imu data

Reply:
@chobitsfan, or we could pass in the imuData that should be used to the CorrectExtNavVelForSensorOffset function.

@priseborough what do you think?  Does it matter if we do the correction using delayed or new IMU data?  The delayed IMU data is not improved in anyway as it ages from "New" to "Delayed"?  maybe that sounds like a funny question but I don't really know what goes on in the EKF.

Reply:
> we could pass in the imuData that should be used to the CorrectExtNavVelForSensorOffset function.

Good idea. It is a more general way.

Reply:
It is important that the offset correction uses IMU and external nav data both at the fusion time horizon which is the delayed data. That is where the data is brought together with a common time stamp and correction for different delays  in each data source.

We should avoid using non tie delayed data inside the EKF as they will not be time aligned with the fusion time horizon.

Reply:
I see. Thank you very much @priseborough . I will fix it

Reply:
I think this is all fixed now so marking as resolved.